--- a/src/gui/painting/qcolor.cpp	2026-08-22 17:50:07
+++ b/src/gui/painting/qcolor.cpp	2026-08-22 17:36:45
@@ -96,10 +96,10 @@
         g = (g << 4) | (g >> 8);
         b = (b << 4) | (b >> 8);
     } else if (len == 8) {
-        a = hex2int(name + 0, 2) * 0x101;
-        r = hex2int(name + 2, 2) * 0x101;
-        g = hex2int(name + 4, 2) * 0x101;
-        b = hex2int(name + 6, 2) * 0x101;
+        r = hex2int(name + 0, 2) * 0x101;
+        g = hex2int(name + 2, 2) * 0x101;
+        b = hex2int(name + 4, 2) * 0x101;
+        a = hex2int(name + 6, 2) * 0x101;
     } else if (len == 6) {
         r = hex2int(name + 0, 2) * 0x101;
         g = hex2int(name + 2, 2) * 0x101;
--- a/src/widgets/graphicsview/qgraphicsitem.cpp	2026-08-22 17:50:07
+++ b/src/widgets/graphicsview/qgraphicsitem.cpp	2026-08-22 17:36:46
@@ -7380,6 +7380,7 @@
                 }
 
                 item->setPos(initialPositions.value(item) + currentParentPos - buttonDownParentPos);
+                item->itemChange( QGraphicsItem::ItemMoved, item->scenePos() );
 
                 if (item->flags() & ItemIsSelectable)
                     item->setSelected(true);
--- a/src/widgets/graphicsview/qgraphicsitem.h	2026-08-22 17:50:07
+++ b/src/widgets/graphicsview/qgraphicsitem.h	2026-08-22 17:36:46
@@ -144,7 +144,8 @@
         ItemScaleChange,
         ItemScaleHasChanged,
         ItemTransformOriginPointChange,
-        ItemTransformOriginPointHasChanged
+        ItemTransformOriginPointHasChanged,
+        ItemMoved
     };
 
     enum CacheMode {
--- a/src/widgets/graphicsview/qgraphicsscene.cpp	2026-08-22 17:50:07
+++ b/src/widgets/graphicsview/qgraphicsscene.cpp	2026-08-22 17:36:46
@@ -1130,16 +1130,15 @@
     if (!view)
         return q->items(scenePos, Qt::IntersectsItemShape, Qt::DescendingOrder, QTransform());
 
-    const QRectF pointRect(QPointF(widget->mapFromGlobal(screenPos)), QSizeF(1, 1));
+    // Sol: we're testing a point so use the pixel intersection test in all cases (prev. would do rect tests?)
+
     if (!view->isTransformed())
-        return q->items(pointRect, Qt::IntersectsItemShape, Qt::DescendingOrder);
+        return q->items(widget->mapFromGlobal(screenPos), Qt::IntersectsItemShape,
+                        Qt::DescendingOrder);
 
     const QTransform viewTransform = view->viewportTransform();
-    if (viewTransform.type() <= QTransform::TxScale) {
-        return q->items(viewTransform.inverted().mapRect(pointRect), Qt::IntersectsItemShape,
-                        Qt::DescendingOrder, viewTransform);
-    }
-    return q->items(viewTransform.inverted().map(pointRect), Qt::IntersectsItemShape,
+    return q->items(viewTransform.inverted().map(widget->mapFromGlobal(screenPos)),
+                    Qt::IntersectsItemShape,
                     Qt::DescendingOrder, viewTransform);
 }
 
--- a/src/widgets/graphicsview/qgraphicssceneindex.cpp	2026-08-22 17:50:07
+++ b/src/widgets/graphicsview/qgraphicssceneindex.cpp	2026-08-22 17:36:46
@@ -432,6 +432,13 @@
 QList<QGraphicsItem *> QGraphicsSceneIndex::items(const QRectF &rect, Qt::ItemSelectionMode mode,
                                                   Qt::SortOrder order, const QTransform &deviceTransform) const
 {
+
+    // Garry: if passing in a rect of 1 pixel size, just use pixel lookup
+    if ( rect.width() == 1 && rect.height() == 1 )
+    {
+        return items( QPoint( rect.left(), rect.top() ), mode, order, deviceTransform );
+    }
+
     Q_D(const QGraphicsSceneIndex);
     QRectF exposeRect = rect;
     _q_adjustRect(&exposeRect);
--- a/src/widgets/graphicsview/qgraphicsview.cpp	2026-08-22 17:50:07
+++ b/src/widgets/graphicsview/qgraphicsview.cpp	2026-08-22 17:36:46
@@ -2306,22 +2306,12 @@
     Q_D(const QGraphicsView);
     if (!d->scene)
         return QList<QGraphicsItem *>();
-    // ### Unify these two, and use the items(QPointF) version in
-    // QGraphicsScene instead. The scene items function could use the viewport
-    // transform to map the point to a rect/polygon.
-    if ((d->identityMatrix || d->matrix.type() <= QTransform::TxScale)) {
-        // Use the rect version
-        QTransform xinv = viewportTransform().inverted();
-        return d->scene->items(xinv.mapRect(QRectF(pos.x(), pos.y(), 1, 1)),
-                               Qt::IntersectsItemShape,
-                               Qt::DescendingOrder,
-                               viewportTransform());
-    }
-    // Use the polygon version
-    return d->scene->items(mapToScene(pos.x(), pos.y(), 1, 1),
-                           Qt::IntersectsItemShape,
-                           Qt::DescendingOrder,
-                           viewportTransform());
+
+    QTransform xinv = viewportTransform().inverted();
+    return d->scene->items(xinv.map(pos),
+                            Qt::IntersectsItemShape,
+                            Qt::DescendingOrder,
+                            viewportTransform());
 }
 
 /*!
--- a/src/widgets/kernel/qwidget.cpp	2026-08-22 17:50:07
+++ b/src/widgets/kernel/qwidget.cpp	2026-08-22 17:36:46
@@ -10561,7 +10561,17 @@
                 q->windowHandle()->setTransientParent(nullptr);
                 q->windowHandle()->setParent(nullptr);
             }
+        }
+        // facepunch - backport of the Qt 6 fix for QTBUG-122747 (qtbase c956eb8eddb1): when an
+        // alien widget is reparented, the QWindows of its created native descendants were left
+        // anchored to the old native parent window, leaving them orphaned and invisible.
+        // Recursively re-anchor them to the closest native window of the new parent chain.
+        else {
+            QWidget *parentWithWindow =
+                newparent ? (newparent->windowHandle() ? newparent : newparent->nativeParentWidget()) : nullptr;
+            reparentWidgetWindowChildren(parentWithWindow);
         }
+        // end facepunch
     }
 
     if (!newparent) {
@@ -10620,6 +10630,48 @@
     }
 }
 
+// facepunch - backport of the Qt 6 reparenting fix (QTBUG-122747, qtbase c956eb8eddb1
+// + windowFlags follow-up 6c036012b5f2), see setParent_sys
+void QWidgetPrivate::reparentWidgetWindows(QWidget *parentWithWindow, Qt::WindowFlags windowFlags)
+{
+    if (QWindow *window = windowHandle()) {
+        // Reparent this QWindow, and all QWindow children will follow
+        if (parentWithWindow) {
+            if (windowFlags & Qt::Window) {
+                // Top level windows can only have transient parents,
+                // and the transient parent must be another top level.
+                QWidget *topLevel = parentWithWindow->window();
+                auto *transientParent = topLevel->windowHandle();
+                Q_ASSERT(transientParent);
+                window->setTransientParent(transientParent);
+                window->setParent(nullptr);
+            } else {
+                auto *parentWindow = parentWithWindow->windowHandle();
+                window->setTransientParent(nullptr);
+                window->setParent(parentWindow);
+            }
+        } else {
+            window->setTransientParent(nullptr);
+            window->setParent(nullptr);
+        }
+    } else {
+        reparentWidgetWindowChildren(parentWithWindow);
+    }
+}
+
+void QWidgetPrivate::reparentWidgetWindowChildren(QWidget *parentWithWindow)
+{
+    for (auto *child : qAsConst(children)) {
+        if (auto *childWidget = qobject_cast<QWidget*>(child)) {
+            auto *childPrivate = QWidgetPrivate::get(childWidget);
+            // Child widgets with QWindows should always continue to be child
+            // windows, so we pass on the child's current window flags here.
+            childPrivate->reparentWidgetWindows(parentWithWindow, childWidget->windowFlags());
+        }
+    }
+}
+// end facepunch
+
 /*!
     Scrolls the widget including its children \a dx pixels to the
     right and \a dy downward. Both \a dx and \a dy may be negative.
@@ -10916,15 +10968,6 @@
     Q_STATIC_ASSERT_X(sizeof(d->high_attributes)*8 >= (Qt::WA_AttributeCount - sizeof(uint)*8),
                       "QWidget::setAttribute(WidgetAttribute, bool): "
                       "QWidgetPrivate::high_attributes[] too small to contain all attributes in WidgetAttribute");
-#ifdef Q_OS_WIN
-    // ### Don't use PaintOnScreen+paintEngine() to do native painting in some future release
-    if (attribute == Qt::WA_PaintOnScreen && on && windowType() != Qt::Desktop && !inherits("QGLWidget")) {
-        // see ::paintEngine for details
-        paintEngine();
-        if (d->noPaintOnScreen)
-            return;
-    }
-#endif
 
     // Don't set WA_NativeWindow on platforms that don't support it -- except for QGLWidget, which depends on it
     if (attribute == Qt::WA_NativeWindow && !d->mustHaveWindowHandle) {
@@ -12848,7 +12891,23 @@
     return debug;
 }
 #endif // !QT_NO_DEBUG_STREAM
+
+// facepunch
+
+void QWidget::Polish()
+{
+    style()->unpolish( this );
+    style()->polish( this );
+
+    QEvent event(QEvent::StyleChange);
+    QApplication::sendEvent( this, &event);
 
+    update();
+    updateGeometry();
+}
+
+// end facepunch
+
 QT_END_NAMESPACE
 
 #include "moc_qwidget.cpp"
--- a/src/widgets/kernel/qwidget.h	2026-08-22 17:50:07
+++ b/src/widgets/kernel/qwidget.h	2026-08-22 17:36:46
@@ -745,6 +745,31 @@
     Q_PRIVATE_SLOT(d_func(), void _q_showIfNotHidden())
 
     QWidgetData *data;
+
+// facepunch changes
+public:
+
+    //
+    // Lets let managed class names appear in stylesheets just like regular classnames
+    // This should make things a bit tidier, since we won't have to do dumb shit like 
+    // name every object and do classes based on that.
+    //
+    QStringList managedClassNames;
+
+    void AddClassName( QString name )
+    {
+        managedClassNames.append( name );
+    }
+
+    void CollectClassNames( QStringList& list )
+    {
+        if ( managedClassNames.count() == 0 )
+            return;
+            
+        list = managedClassNames + list;
+    }
+
+    void Polish();
 };
 
 Q_DECLARE_OPERATORS_FOR_FLAGS(QWidget::RenderFlags)
--- a/src/widgets/kernel/qwidget_p.h	2026-08-22 17:50:07
+++ b/src/widgets/kernel/qwidget_p.h	2026-08-22 17:36:46
@@ -403,6 +403,10 @@
     void showChildren(bool spontaneous);
     void hideChildren(bool spontaneous);
     void setParent_sys(QWidget *parent, Qt::WindowFlags);
+    // facepunch - backport of the Qt 6 reparenting fix for native descendant windows
+    void reparentWidgetWindows(QWidget *parentWithWindow, Qt::WindowFlags windowFlags = {});
+    void reparentWidgetWindowChildren(QWidget *parentWithWindow);
+    // end facepunch
     void scroll_sys(int dx, int dy);
     void scroll_sys(int dx, int dy, const QRect &r);
     void deactivateWidgetCleanup();
--- a/src/widgets/styles/qstylesheetstyle.cpp	2026-08-22 17:50:07
+++ b/src/widgets/styles/qstylesheetstyle.cpp	2026-08-22 17:36:46
@@ -1507,7 +1507,10 @@
     {
         if (isNullNode(node))
             return QStringList();
-        const QMetaObject *metaObject = OBJECT_PTR(node)->metaObject();
+
+        QObject *obj = OBJECT_PTR(node);
+
+        const QMetaObject *metaObject = obj->metaObject();
 #ifndef QT_NO_TOOLTIP
         if (qstrcmp(metaObject->className(), "QTipLabel") == 0)
             return QStringList(QLatin1String("QToolTip"));
@@ -1517,6 +1520,13 @@
             result += QString::fromLatin1(metaObject->className()).replace(QLatin1Char(':'), QLatin1Char('-'));
             metaObject = metaObject->superClass();
         } while (metaObject != nullptr);
+
+        QWidget *w = qobject_cast<QWidget *>(obj);
+        if ( w )
+        {
+            w->CollectClassNames( result );
+        }
+
         return result;
     }
     QString attribute(NodePtr node, const QString& name) const override
@@ -1559,7 +1569,8 @@
     {
         if (isNullNode(node))
             return false;
-        const QMetaObject *metaObject = OBJECT_PTR(node)->metaObject();
+        QObject *obj = OBJECT_PTR(node);
+        const QMetaObject *metaObject = obj->metaObject();
 #ifndef QT_NO_TOOLTIP
         if (qstrcmp(metaObject->className(), "QTipLabel") == 0)
             return nodeName == QLatin1String("QToolTip");
@@ -1576,6 +1587,13 @@
                 return true;
             metaObject = metaObject->superClass();
         } while (metaObject != nullptr);
+
+        QWidget *w = qobject_cast<QWidget *>(obj);
+        if ( w )
+        {
+            return w->managedClassNames.contains( nodeName );
+        }
+
         return false;
     }
     bool hasAttributes(NodePtr) const override
--- a/src/widgets/widgets/qabstractscrollarea.cpp	2026-08-22 17:50:07
+++ b/src/widgets/widgets/qabstractscrollarea.cpp	2026-08-22 17:36:46
@@ -354,29 +354,11 @@
 
     const bool hasCornerWidget = (cornerWidget != nullptr);
 
-    QPoint cornerOffset((needv && vscrollOverlap == 0) ? vsbExt : 0, (needh && hscrollOverlap == 0) ? hsbExt : 0);
-    QRect controlsRect;
-    QRect viewportRect;
+    QPoint cornerOffset( 0, 0 );
+    QRect controlsRect = widgetRect;
+    QRect viewportRect = widgetRect;
 
-    // In FrameOnlyAroundContents mode the frame is drawn between the controls and
-    // the viewport, else the frame rect is equal to the widget rect.
-    if ((frameStyle != QFrame::NoFrame) &&
-        q->style()->styleHint(QStyle::SH_ScrollView_FrameOnlyAroundContents, &opt, q)) {
-        controlsRect = widgetRect;
-        const int spacing = q->style()->pixelMetric(QStyle::PM_ScrollView_ScrollBarSpacing, &opt, q);
-        const QPoint cornerExtra(needv ? spacing + vscrollOverlap : 0, needh ? spacing + hscrollOverlap : 0);
-        QRect frameRect = widgetRect;
-        frameRect.adjust(0, 0, -cornerOffset.x() - cornerExtra.x(), -cornerOffset.y() - cornerExtra.y());
-        q->setFrameRect(QStyle::visualRect(opt.direction, opt.rect, frameRect));
-        // The frame rect needs to be in logical coords, however we need to flip
-        // the contentsRect back before passing it on to the viewportRect
-        // since the viewportRect has its logical coords calculated later.
-        viewportRect = QStyle::visualRect(opt.direction, opt.rect, q->contentsRect());
-    } else {
-        q->setFrameRect(QStyle::visualRect(opt.direction, opt.rect, widgetRect));
-        controlsRect = q->contentsRect();
-        viewportRect = QRect(controlsRect.topLeft(), controlsRect.bottomRight() - cornerOffset);
-    }
+    q->setFrameRect(QStyle::visualRect(opt.direction, opt.rect, widgetRect));
 
     cornerOffset = QPoint(needv ? vsbExt : 0, needh ? hsbExt : 0);
 
@@ -430,7 +412,8 @@
         scrollBarContainers[Qt::Vertical]->raise();
     }
 
-    if (cornerWidget) {
+    if (cornerWidget) 
+    {
         const QRect cornerWidgetRect(cornerPoint, controlsRect.bottomRight());
         cornerWidget->setGeometry(QStyle::visualRect(opt.direction, opt.rect, cornerWidgetRect));
     }
--- a/src/widgets/widgets/qcombobox.cpp	2026-08-22 17:50:07
+++ b/src/widgets/widgets/qcombobox.cpp	2026-08-22 17:36:46
@@ -3473,6 +3473,9 @@
 #if QT_CONFIG(wheelevent)
 void QComboBox::wheelEvent(QWheelEvent *e)
 {
+    // garry: fuck this behaviour
+    return;
+
     Q_D(QComboBox);
     QStyleOptionComboBox opt;
     initStyleOption(&opt);
--- a/src/widgets/widgets/qdockarealayout.cpp	2026-08-22 17:50:07
+++ b/src/widgets/widgets/qdockarealayout.cpp	2026-08-22 17:36:46
@@ -139,7 +139,7 @@
 QSize QDockAreaLayoutItem::minimumSize() const
 {
     if (widgetItem)
-        return widgetItem->minimumSize().grownBy(widgetItem->widget()->contentsMargins());
+        return widgetItem->minimumSize().grownBy(widgetItem->widget()->contentsMargins()).grownBy( { 4, 4, 4, 4 } );
     if (subinfo != nullptr)
         return subinfo->minimumSize();
     return QSize(0, 0);
@@ -174,10 +174,13 @@
 {
     if (placeHolderItem != nullptr)
         return QSize(0, 0);
+
     if (widgetItem)
-        return widgetItem->sizeHint().grownBy(widgetItem->widget()->contentsMargins());
+        return widgetItem->sizeHint().grownBy(widgetItem->widget()->contentsMargins()).grownBy( { 4, 4, 4, 4 } );
+
     if (subinfo != nullptr)
         return subinfo->sizeHint();
+
     return QSize(-1, -1);
 }
 
@@ -979,26 +982,26 @@
 void QDockAreaLayoutInfo::unnest(int index)
 {
     QDockAreaLayoutItem &item = item_list[index];
-    if (item.subinfo == nullptr)
-        return;
-    if (item.subinfo->item_list.count() > 1)
-        return;
 
-    if (item.subinfo->item_list.count() == 0) {
+    if ( item.subinfo == nullptr ) return;
+
+    if (item.subinfo->item_list.count() == 0) 
+    {
         item_list.removeAt(index);
-    } else if (item.subinfo->item_list.count() == 1) {
-        QDockAreaLayoutItem &child = item.subinfo->item_list.first();
-        if (child.widgetItem != nullptr) {
-            item.widgetItem = child.widgetItem;
-            delete item.subinfo;
-            item.subinfo = nullptr;
-        } else if (child.subinfo != nullptr) {
-            QDockAreaLayoutInfo *tmp = item.subinfo;
-            item.subinfo = child.subinfo;
-            child.subinfo = nullptr;
-            tmp->item_list.clear();
-            delete tmp;
-        }
+        return;
+    } 
+
+    if ( item.subinfo->item_list.count() > 1 ) return;
+
+    QDockAreaLayoutItem &child = item.subinfo->item_list.first();
+
+    if (child.subinfo != nullptr) 
+    {
+        QDockAreaLayoutInfo *tmp = item.subinfo;
+        item.subinfo = child.subinfo;
+        child.subinfo = nullptr;
+        tmp->item_list.clear();
+        delete tmp;
     }
 }
 
@@ -1037,7 +1040,32 @@
     Q_ASSERT(item.widgetItem != nullptr);
     Q_ASSERT(item.flags & QDockAreaLayoutItem::GapItem);
     item.flags &= ~QDockAreaLayoutItem::GapItem;
-    return item.widgetItem;
+
+    //
+    // garry: if we dropped onto a tabbed area, that's cool - just carry on
+    //
+    if ( tabbed )
+        return item.widgetItem;
+
+    //
+    // If we dropped somewhere that isn't a tabbed area, we want to make a new layout
+    // info to enable tabs for it. We want tabs even if there's only one tab.
+    //
+
+    auto widgetItem = item.widgetItem;
+    QDockAreaLayoutInfo *new_info = new QDockAreaLayoutInfo(sep, dockPos, o, tabBarShape, mainWindow);
+    new_info->item_list.append(QDockAreaLayoutItem(widgetItem));
+
+    item.widgetItem = nullptr;
+    item.subinfo = new_info;
+
+    new_info->tabbed = true;
+    new_info->updateTabBar();
+    new_info->setCurrentTab( widgetItem->widget() );
+
+    // qDebug() << "QDockAreaLayoutInfo::plug " << tabbed;
+
+    return widgetItem;
 }
 
 QLayoutItem *QDockAreaLayoutInfo::unplug(const QList<int> &path)
@@ -1115,7 +1143,7 @@
     QDockWidgetLayout *layout
         = qobject_cast<QDockWidgetLayout*>(widget->layout());
     if (layout && layout->nativeWindowDeco())
-        titleHeight = layout->titleHeight();
+        titleHeight = 0;
 
     QRect result = widget->geometry();
     result.adjust(0, -titleHeight, 0, 0);
@@ -1690,21 +1718,36 @@
 }
 #endif // QT_CONFIG(tabbar)
 
-void QDockAreaLayoutInfo::split(int index, Qt::Orientation orientation,
-                                       QLayoutItem *dockWidgetItem)
+void QDockAreaLayoutInfo::split( int index, Qt::Orientation orientation, QLayoutItem *dockWidgetItem )
 {
-    if (orientation == o) {
+   // qDebug() << "splitting" << index;
+
+    if (orientation == o) 
+    {
+      //  qDebug() << "orientation matches" << orientation;
+
         item_list.insert(index + 1, QDockAreaLayoutItem(dockWidgetItem));
-    } else {
-#if !QT_CONFIG(tabbar)
-        const int tabBarShape = 0;
-#endif
-        QDockAreaLayoutInfo *new_info
-            = new QDockAreaLayoutInfo(sep, dockPos, orientation, tabBarShape, mainWindow);
-        item_list[index].subinfo = new_info;
-        new_info->item_list.append(QDockAreaLayoutItem(item_list.at(index).widgetItem));
-        item_list[index].widgetItem = nullptr;
+    } 
+    else 
+    {
+       // qDebug() << "orientation no match" << orientation;
+        //qDebug() << "subinfo is null? " << (item_list[index].subinfo == nullptr);
+        QDockAreaLayoutInfo *new_info = new QDockAreaLayoutInfo(sep, dockPos, orientation, tabBarShape, mainWindow);
+        
+        if ( item_list[index].subinfo != nullptr )
+        {
+            new_info->item_list.append( QDockAreaLayoutItem( item_list[index].subinfo ) );
+            item_list[index].subinfo = nullptr;
+        }
+        else
+        {
+            new_info->item_list.append(QDockAreaLayoutItem(item_list.at(index).widgetItem));
+            item_list[index].widgetItem = nullptr;
+        }
+
         new_info->item_list.append(QDockAreaLayoutItem(dockWidgetItem));
+
+        item_list[index].subinfo = new_info;
     }
 }
 
@@ -2084,7 +2127,6 @@
     if (that->tabBar == nullptr) {
         that->tabBar = mainWindowLayout()->getTabBar();
         that->tabBar->setShape(static_cast<QTabBar::Shape>(tabBarShape));
-        that->tabBar->setDrawBase(true);
     }
 
     const QSignalBlocker blocker(tabBar);
@@ -2107,32 +2149,33 @@
         QDockWidget *dw = qobject_cast<QDockWidget*>(item.widgetItem->widget());
         QString title = dw->d_func()->fixedWindowTitle;
         quintptr id = tabId(item);
-        if (tab_idx == tabBar->count()) {
+        
+        if (tab_idx == tabBar->count()) 
+        {
             tabBar->insertTab(tab_idx, title);
-#ifndef QT_NO_TOOLTIP
             tabBar->setTabToolTip(tab_idx, title);
-#endif
             tabBar->setTabData(tab_idx, id);
-        } else if (qvariant_cast<quintptr>(tabBar->tabData(tab_idx)) != id) {
+        } 
+        else if (qvariant_cast<quintptr>(tabBar->tabData(tab_idx)) != id) 
+        {
             if (tab_idx + 1 < tabBar->count()
                     && qvariant_cast<quintptr>(tabBar->tabData(tab_idx + 1)) == id)
                 tabBar->removeTab(tab_idx);
             else {
                 tabBar->insertTab(tab_idx, title);
-#ifndef QT_NO_TOOLTIP
                 tabBar->setTabToolTip(tab_idx, title);
-#endif
                 tabBar->setTabData(tab_idx, id);
             }
         }
 
-        if (title != tabBar->tabText(tab_idx)) {
+        if (title != tabBar->tabText(tab_idx)) 
+        {
             tabBar->setTabText(tab_idx, title);
-#ifndef QT_NO_TOOLTIP
             tabBar->setTabToolTip(tab_idx, title);
-#endif
         }
 
+        dw->initializeTabButton( tabBar, tab_idx );
+
         ++tab_idx;
     }
 
@@ -2147,7 +2190,7 @@
         dwgw->adjustFlags();
 
     //returns if the tabbar is visible or not
-    return ( (gap ? 1 : 0) + tabBar->count()) > 1;
+    return ( (gap ? 1 : 0) + tabBar->count()) > 0;
 }
 
 void QDockAreaLayoutInfo::setTabBarShape(int shape)
@@ -3075,33 +3118,33 @@
     return true;
 }
 
-void QDockAreaLayout::addDockWidget(QInternal::DockPosition pos, QDockWidget *dockWidget,
-                                             Qt::Orientation orientation)
+void QDockAreaLayout::addDockWidget(QInternal::DockPosition pos, QDockWidget *dockWidget, Qt::Orientation orientation)
 {
     QLayoutItem *dockWidgetItem = new QDockWidgetItem(dockWidget);
     QDockAreaLayoutInfo &info = docks[pos];
-    if (orientation == info.o || info.item_list.count() <= 1) {
-        // empty dock areas, or dock areas containing exactly one widget can have their orientation
-        // switched.
-        info.o = orientation;
 
-        QDockAreaLayoutItem new_item(dockWidgetItem);
-        info.item_list.append(new_item);
-#if QT_CONFIG(tabbar)
-        if (info.tabbed && !new_item.skip()) {
-            info.updateTabBar();
-            info.setCurrentTabId(tabId(new_item));
-        }
-#endif
-    } else {
-#if QT_CONFIG(tabbar)
-        int tbshape = info.tabBarShape;
-#else
-        int tbshape = 0;
-#endif
-        QDockAreaLayoutInfo new_info(&sep, pos, orientation, tbshape, mainWindow);
+    QDockAreaLayoutInfo* new_tab = new QDockAreaLayoutInfo( &sep, pos, orientation, info.tabBarShape, mainWindow );
+    new_tab->item_list.append( QDockAreaLayoutItem(dockWidgetItem) );
+    new_tab->tabbed = true;
+    new_tab->updateTabBar();
+    new_tab->setCurrentTabId(0);
+
+    if ( info.item_list.count() == 0 )
+    {
+        //
+        // garry: if we're the only item, wrap it so we can have a tab
+        //
+        info.item_list.append( QDockAreaLayoutItem( new_tab ) );
+    } 
+    else 
+    {
+        //
+        // garry: If we're adding more than one, split us up
+        //
+        QDockAreaLayoutInfo new_info(&sep, pos, orientation, info.tabBarShape, mainWindow);
         new_info.item_list.append(QDockAreaLayoutItem(new QDockAreaLayoutInfo(info)));
-        new_info.item_list.append(QDockAreaLayoutItem(dockWidgetItem));
+        new_info.item_list.append(QDockAreaLayoutItem(new_tab));
+
         info = new_info;
     }
 
@@ -3172,17 +3215,31 @@
     }
 }
 
-void QDockAreaLayout::splitDockWidget(QDockWidget *after,
-                                               QDockWidget *dockWidget,
-                                               Qt::Orientation orientation)
+void QDockAreaLayout::splitDockWidget(QDockWidget *after, QDockWidget *dockWidget, Qt::Orientation orientation)
 {
-    const QList<int> path = indexOf(after);
+    QList<int> path = indexOf(after);
     if (path.isEmpty())
         return;
 
+    int targetId = path.last();
     QDockAreaLayoutInfo *info = this->info(path);
+
+    //
+    // Garry: If our target is tabbed, we can't just split it up. We need
+    // to add to its parent instead. This is broken in default qt too. There's
+    // forum posts and everything.
+    //
+    if ( info->tabbed )
+    {
+        path.removeAt( path.size() - 1 );
+        QDockAreaLayoutInfo *parentInfo = this->info(path);
+
+        targetId = path.last();
+        info = parentInfo;
+    }
+
     Q_ASSERT(info != nullptr);
-    info->split(path.last(), orientation, new QDockWidgetItem(dockWidget));
+    info->split(targetId, orientation, new QDockWidgetItem(dockWidget));
 
     removePlaceHolder(dockWidget->objectName());
 }
--- a/src/widgets/widgets/qdockwidget.cpp	2026-08-22 17:50:07
+++ b/src/widgets/widgets/qdockwidget.cpp	2026-08-22 17:36:46
@@ -250,8 +250,7 @@
 ** QDockWidgetLayout
 */
 
-QDockWidgetLayout::QDockWidgetLayout(QWidget *parent)
-    : QLayout(parent), verticalTitleBar(false), item_list(RoleCount, 0)
+QDockWidgetLayout::QDockWidgetLayout(QWidget *parent) : QLayout(parent), verticalTitleBar(false), item_list(RoleCount, 0)
 {
 }
 
@@ -296,7 +295,7 @@
  */
 bool QDockWidgetLayout::nativeWindowDeco(bool floating) const
 {
-    return wmSupportsNativeWindowDeco() && floating && item_list.at(QDockWidgetLayout::TitleBar) == 0;
+    return wmSupportsNativeWindowDeco() && floating;
 }
 
 
@@ -350,14 +349,6 @@
 {
     QSize result = content;
 
-    if (verticalTitleBar) {
-        result.setHeight(qMax(result.height(), minimumTitleWidth()));
-        result.setWidth(qMax(content.width(), 0));
-    } else {
-        result.setHeight(qMax(result.height(), 0));
-        result.setWidth(qMax(content.width(), minimumTitleWidth()));
-    }
-
     QDockWidget *w = qobject_cast<QDockWidget*>(parentWidget());
     const bool nativeDeco = nativeWindowDeco(floating);
 
@@ -365,7 +356,7 @@
             ? w->style()->pixelMetric(QStyle::PM_DockWidgetFrameWidth, nullptr, w)
             : 0;
 
-    const int th = titleHeight();
+    const int th = 0;
     if (!nativeDeco) {
         if (verticalTitleBar)
             result += QSize(th + 2*fw, 2*fw);
@@ -483,122 +474,26 @@
 static inline int perp(bool vertical, const QSize &size)
 {
     return vertical ? size.width() : size.height();
-}
-
-int QDockWidgetLayout::minimumTitleWidth() const
-{
-    QDockWidget *q = qobject_cast<QDockWidget*>(parentWidget());
-
-    if (QWidget *title = widgetForRole(TitleBar))
-        return pick(verticalTitleBar, title->minimumSizeHint());
-
-    QSize closeSize(0, 0);
-    QSize floatSize(0, 0);
-    if (hasFeature(q, QDockWidget::DockWidgetClosable)) {
-        if (QLayoutItem *item = item_list[CloseButton])
-            closeSize = item->widget()->sizeHint();
-    }
-    if (hasFeature(q, QDockWidget::DockWidgetFloatable)) {
-        if (QLayoutItem *item = item_list[FloatButton])
-            floatSize = item->widget()->sizeHint();
-    }
-
-    int titleHeight = this->titleHeight();
-
-    int mw = q->style()->pixelMetric(QStyle::PM_DockWidgetTitleMargin, nullptr, q);
-    int fw = q->style()->pixelMetric(QStyle::PM_DockWidgetFrameWidth, nullptr, q);
-
-    return pick(verticalTitleBar, closeSize)
-            + pick(verticalTitleBar, floatSize)
-            + titleHeight + 2*fw + 3*mw;
 }
 
-int QDockWidgetLayout::titleHeight() const
-{
-    QDockWidget *q = qobject_cast<QDockWidget*>(parentWidget());
-
-    if (QWidget *title = widgetForRole(TitleBar))
-        return perp(verticalTitleBar, title->sizeHint());
-
-    QSize closeSize(0, 0);
-    QSize floatSize(0, 0);
-    if (QLayoutItem *item = item_list[CloseButton])
-        closeSize = item->widget()->sizeHint();
-    if (QLayoutItem *item = item_list[FloatButton])
-        floatSize = item->widget()->sizeHint();
-
-    int buttonHeight = qMax(perp(verticalTitleBar, closeSize),
-                            perp(verticalTitleBar, floatSize));
-
-    QFontMetrics titleFontMetrics = q->fontMetrics();
-    int mw = q->style()->pixelMetric(QStyle::PM_DockWidgetTitleMargin, nullptr, q);
-
-    return qMax(buttonHeight + 2, titleFontMetrics.height() + 2*mw);
-}
-
 void QDockWidgetLayout::setGeometry(const QRect &geometry)
 {
-    QDockWidget *q = qobject_cast<QDockWidget*>(parentWidget());
+    QRect rect = geometry;
 
-    bool nativeDeco = nativeWindowDeco();
+    //
+    // I want padding inside my dock panel but I don't
+    // know how to access the padding from the stylesheet
+    // so I'm hard coding this here cuz fuck it, every other
+    // thing in this framework is hardcoded like this.
+    //
+    rect.setLeft( rect.left() + 1 );
+    rect.setTop( rect.top() + 1 );
+    rect.setRight( rect.right() - 1 );
+    rect.setBottom( rect.bottom() - 1 );
 
-    int fw = q->isFloating() && !nativeDeco
-            ? q->style()->pixelMetric(QStyle::PM_DockWidgetFrameWidth, nullptr, q)
-            : 0;
+    if (QLayoutItem *item = item_list[Content])
+        item->setGeometry(rect);
 
-    if (nativeDeco) {
-        if (QLayoutItem *item = item_list[Content])
-            item->setGeometry(geometry);
-    } else {
-        int titleHeight = this->titleHeight();
-
-        if (verticalTitleBar) {
-            _titleArea = QRect(QPoint(fw, fw),
-                                QSize(titleHeight, geometry.height() - (fw * 2)));
-        } else {
-            _titleArea = QRect(QPoint(fw, fw),
-                                QSize(geometry.width() - (fw * 2), titleHeight));
-        }
-
-        if (QLayoutItem *item = item_list[TitleBar]) {
-            item->setGeometry(_titleArea);
-        } else {
-            QStyleOptionDockWidget opt;
-            q->initStyleOption(&opt);
-
-            if (QLayoutItem *item = item_list[CloseButton]) {
-                if (!item->isEmpty()) {
-                    QRect r = q->style()
-                        ->subElementRect(QStyle::SE_DockWidgetCloseButton,
-                                            &opt, q);
-                    if (!r.isNull())
-                        item->setGeometry(r);
-                }
-            }
-
-            if (QLayoutItem *item = item_list[FloatButton]) {
-                if (!item->isEmpty()) {
-                    QRect r = q->style()
-                        ->subElementRect(QStyle::SE_DockWidgetFloatButton,
-                                            &opt, q);
-                    if (!r.isNull())
-                        item->setGeometry(r);
-                }
-            }
-        }
-
-        if (QLayoutItem *item = item_list[Content]) {
-            QRect r = geometry;
-            if (verticalTitleBar) {
-                r.setLeft(_titleArea.right() + 1);
-                r.adjust(0, fw, -fw, -fw);
-            } else {
-                r.setTop(_titleArea.bottom() + 1);
-                r.adjust(fw, 0, -fw, -fw);
-            }
-            item->setGeometry(r);
-        }
-    }
 }
 
 void QDockWidgetLayout::setVerticalTitleBar(bool b)
@@ -657,16 +552,6 @@
     QDockWidgetLayout *layout = new QDockWidgetLayout(q);
     layout->setSizeConstraint(QLayout::SetMinAndMaxSize);
 
-    QAbstractButton *button = new QDockWidgetTitleButton(q);
-    button->setObjectName(QLatin1String("qt_dockwidget_floatbutton"));
-    QObject::connect(button, SIGNAL(clicked()), q, SLOT(_q_toggleTopLevel()));
-    layout->setWidgetForRole(QDockWidgetLayout::FloatButton, button);
-
-    button = new QDockWidgetTitleButton(q);
-    button->setObjectName(QLatin1String("qt_dockwidget_closebutton"));
-    QObject::connect(button, SIGNAL(clicked()), q, SLOT(close()));
-    layout->setWidgetForRole(QDockWidgetLayout::CloseButton, button);
-
     font = QApplication::font("QDockWidgetTitle");
 
 #ifndef QT_NO_ACTION
@@ -695,6 +580,7 @@
 
     if (!option)
         return;
+
     QDockWidgetLayout *dwlayout = qobject_cast<QDockWidgetLayout*>(layout());
 
     QDockWidgetGroupWindow *floatingTab = qobject_cast<QDockWidgetGroupWindow*>(parent());
@@ -724,39 +610,7 @@
 
 void QDockWidgetPrivate::updateButtons()
 {
-    Q_Q(QDockWidget);
-    QDockWidgetLayout *dwLayout = qobject_cast<QDockWidgetLayout*>(layout);
 
-    QStyleOptionDockWidget opt;
-    q->initStyleOption(&opt);
-
-    bool customTitleBar = dwLayout->widgetForRole(QDockWidgetLayout::TitleBar) != nullptr;
-    bool nativeDeco = dwLayout->nativeWindowDeco();
-    bool hideButtons = nativeDeco || customTitleBar;
-
-    bool canClose = hasFeature(this, QDockWidget::DockWidgetClosable);
-    bool canFloat = hasFeature(this, QDockWidget::DockWidgetFloatable);
-
-    QAbstractButton *button
-        = qobject_cast<QAbstractButton*>(dwLayout->widgetForRole(QDockWidgetLayout::FloatButton));
-    button->setIcon(q->style()->standardIcon(QStyle::SP_TitleBarNormalButton, &opt, q));
-    button->setVisible(canFloat && !hideButtons);
-#ifndef QT_NO_ACCESSIBILITY
-    //: Accessible name for button undocking a dock widget (floating state)
-    button->setAccessibleName(QDockWidget::tr("Float"));
-    button->setAccessibleDescription(QDockWidget::tr("Undocks and re-attaches the dock widget"));
-#endif
-    button
-        = qobject_cast <QAbstractButton*>(dwLayout->widgetForRole(QDockWidgetLayout::CloseButton));
-    button->setIcon(q->style()->standardIcon(QStyle::SP_TitleBarCloseButton, &opt, q));
-    button->setVisible(canClose && !hideButtons);
-#ifndef QT_NO_ACCESSIBILITY
-    //: Accessible name for button closing a dock widget
-    button->setAccessibleName(QDockWidget::tr("Close"));
-    button->setAccessibleDescription(QDockWidget::tr("Closes the dock widget"));
-#endif
-
-    layout->invalidate();
 }
 
 void QDockWidgetPrivate::_q_toggleTopLevel()
@@ -1109,7 +963,7 @@
     r.moveTopLeft(q->mapToGlobal(QPoint(0, 0)));
     QDockWidgetLayout *dwLayout = qobject_cast<QDockWidgetLayout*>(layout);
     if (dwLayout->nativeWindowDeco(true))
-        r.adjust(0, dwLayout->titleHeight(), 0, 0);
+        r.adjust(0, 0, 0, 0);
     setWindowState(true, true, r);
 }
 
@@ -1473,34 +1327,11 @@
 void QDockWidget::paintEvent(QPaintEvent *event)
 {
     Q_UNUSED(event)
-    Q_D(QDockWidget);
 
-    QDockWidgetLayout *layout
-        = qobject_cast<QDockWidgetLayout*>(this->layout());
-    bool customTitleBar = layout->widgetForRole(QDockWidgetLayout::TitleBar) != nullptr;
-    bool nativeDeco = layout->nativeWindowDeco();
-
-    if (!nativeDeco && !customTitleBar) {
-        QStylePainter p(this);
-        // ### Add PixelMetric to change spacers, so style may show border
-        // when not floating.
-        if (isFloating()) {
-            QStyleOptionFrame framOpt;
-            framOpt.init(this);
-            p.drawPrimitive(QStyle::PE_FrameDockWidget, framOpt);
-        }
-
-        // Title must be painted after the frame, since the areas overlap, and
-        // the title may wish to extend out to all sides (eg. Vista style)
-        QStyleOptionDockWidget titleOpt;
-        initStyleOption(&titleOpt);
-        if (font() == QApplication::font("QDockWidget")) {
-            titleOpt.fontMetrics = QFontMetrics(d->font);
-            p.setFont(d->font);
-        }
-
-        p.drawControl(QStyle::CE_DockWidgetTitle, titleOpt);
-    }
+    QStyleOption opt;
+    opt.init(this);
+    QPainter p(this);
+    style()->drawPrimitive(QStyle::PE_Widget, &opt, &p, this);
 }
 
 /*! \reimp */
@@ -1709,32 +1540,38 @@
     \sa titleBarWidget(), DockWidgetVerticalTitleBar
 */
 
-void QDockWidget::setTitleBarWidget(QWidget *widget)
+void QDockWidget::initializeTabButton( QTabBar* tabBar, int index )
 {
-    Q_D(QDockWidget);
-    QDockWidgetLayout *layout
-        = qobject_cast<QDockWidgetLayout*>(this->layout());
-    layout->setWidgetForRole(QDockWidgetLayout::TitleBar, widget);
-    d->updateButtons();
-    if (isWindow()) {
-        //this ensures the native decoration is drawn
-        d->setWindowState(true /*floating*/, true /*unplug*/);
+    //
+    // Things got a bit more complicated here than I'd hoped. This function can get called multiple
+    // times so we need to make sure that we're only doing our thing once. I think it gets called
+    // if you set a tab button. I could make this neater for people overriding it but it's dinner time.
+    //
+
+    auto btn = tabBar->tabButton( index, QTabBar::ButtonPosition::RightSide );
+    if ( btn != nullptr )
+    {
+        const QVariant v = btn->property( "__ownerDock" );
+        if ( v.isValid() && qvariant_cast<QDockWidget*>( v ) == this )
+            return;
     }
-}
 
-/*!
-    \since 4.3
-    Returns the custom title bar widget set on the QDockWidget, or
-    \nullptr if no custom title bar has been set.
+    if ( !(features() & DockWidgetClosable) ) return;
 
-    \sa setTitleBarWidget()
-*/
+    QAbstractButton *closeButton = new QToolButton( nullptr );
+    closeButton->setText( "x" );
+    closeButton->setObjectName( "CloseButton" );
+    closeButton->setProperty( "__ownerDock", QVariant::fromValue( this ) );
+    connect(closeButton, SIGNAL(clicked()), this, SLOT(close()));
 
-QWidget *QDockWidget::titleBarWidget() const
-{
-    QDockWidgetLayout *layout
-        = qobject_cast<QDockWidgetLayout*>(this->layout());
-    return layout->widgetForRole(QDockWidgetLayout::TitleBar);
+    tabBar->setTabButton( index, QTabBar::ButtonPosition::RightSide, closeButton );
+
+    if ( !m_icon.isNull() )
+    {
+        tabBar->setTabIcon( index, m_icon );
+    }
+
+    closeButton->setVisible( false );
 }
 
 QT_END_NAMESPACE
--- a/src/widgets/widgets/qdockwidget.h	2026-08-22 17:50:07
+++ b/src/widgets/widgets/qdockwidget.h	2026-08-22 17:36:46
@@ -51,6 +51,7 @@
 class QDockWidgetPrivate;
 class QMainWindow;
 class QStyleOptionDockWidget;
+class QTabBar;
 
 class Q_WIDGETS_EXPORT QDockWidget : public QWidget
 {
@@ -98,8 +99,13 @@
     void setAllowedAreas(Qt::DockWidgetAreas areas);
     Qt::DockWidgetAreas allowedAreas() const;
 
-    void setTitleBarWidget(QWidget *widget);
-    QWidget *titleBarWidget() const;
+    //
+    // garry: these aren't used anymore
+    //
+    // s&box: kept outside QT_DEPRECATED_SINCE - that evaluates false by default in 5.15,
+    // and qaccessiblewidgets.cpp still calls titleBarWidget() unconditionally.
+    void setTitleBarWidget(QWidget *){}
+    QWidget *titleBarWidget() const{ return nullptr; }
 
     inline bool isAreaAllowed(Qt::DockWidgetArea area) const
     { return (allowedAreas() & area) == area; }
@@ -132,6 +138,16 @@
     friend class QMainWindowLayout;
     friend class QDockWidgetLayout;
     friend class QDockAreaLayoutInfo;
+
+public:
+
+    virtual void initializeTabButton( QTabBar* tabBar, int index );
+    virtual QIcon icon() const{  return m_icon; }
+    virtual void setIcon( const QIcon &icon ) { m_icon = icon; }
+
+    protected:
+
+    QIcon m_icon;
 };
 
 Q_DECLARE_OPERATORS_FOR_FLAGS(QDockWidget::DockWidgetFeatures)
--- a/src/widgets/widgets/qdockwidget_p.h	2026-08-22 17:50:07
+++ b/src/widgets/widgets/qdockwidget_p.h	2026-08-22 17:36:46
@@ -153,15 +153,13 @@
 
     void setGeometry(const QRect &r) override;
 
-    enum Role { Content, CloseButton, FloatButton, TitleBar, RoleCount };
+    enum Role { Content, RoleCount };
     QWidget *widgetForRole(Role r) const;
     void setWidgetForRole(Role r, QWidget *w);
     QLayoutItem *itemForRole(Role r) const;
 
     QRect titleArea() const { return _titleArea; }
 
-    int minimumTitleWidth() const;
-    int titleHeight() const;
     void updateMaxSize();
     static bool wmSupportsNativeWindowDeco();
     bool nativeWindowDeco() const;
--- a/src/widgets/widgets/qmainwindow.h	2026-08-22 17:50:07
+++ b/src/widgets/widgets/qmainwindow.h	2026-08-22 17:36:46
@@ -123,8 +123,8 @@
     bool isSeparator(const QPoint &pos) const;
 
 #if QT_CONFIG(menubar)
-    QMenuBar *menuBar() const;
-    void setMenuBar(QMenuBar *menubar);
+    virtual QMenuBar *menuBar() const;
+    virtual void setMenuBar(QMenuBar *menubar);
 
     QWidget  *menuWidget() const;
     void setMenuWidget(QWidget *menubar);
--- a/src/widgets/widgets/qmainwindowlayout.cpp	2026-08-22 17:50:07
+++ b/src/widgets/widgets/qmainwindowlayout.cpp	2026-08-22 17:36:46
@@ -705,7 +705,7 @@
 #endif // QT_CONFIG(toolbar)
 
 #if QT_CONFIG(dockwidget)
-    dockAreaLayout.rect = r;
+    dockAreaLayout.rect = r.marginsAdded( { -4, 0, -4, -4 } ); // garry: margin on left/right of window
     dockAreaLayout.fitLayout();
 #else
     centralWidgetRect = r;
@@ -1693,57 +1693,95 @@
     bool event(QEvent *e) override;
     void mouseReleaseEvent(QMouseEvent*) override;
     void mouseMoveEvent(QMouseEvent*) override;
+    void leaveEvent(QEvent *event) override;
+    int hovered_index = -1;
 
 };
 
-QMainWindowTabBar::QMainWindowTabBar(QMainWindow *parent)
-    : QTabBar(parent), mainWindow(parent)
+QMainWindowTabBar::QMainWindowTabBar(QMainWindow *parent) : QTabBar(parent), mainWindow(parent)
 {
     setExpanding(false);
+    setMouseTracking(true);
 }
 
+void QMainWindowTabBar::leaveEvent(QEvent *e)
+{
+    if ( hovered_index >= 0 )
+    {
+        auto btnA = tabButton( hovered_index, QTabBar::RightSide );
+        if ( btnA ) btnA->setVisible( false );
+        hovered_index = -1;
+    }
+
+    QTabBar::leaveEvent(e);
+}
+
 void QMainWindowTabBar::mouseMoveEvent(QMouseEvent *e)
 {
-    // The QTabBar handles the moving (reordering) of tabs.
-    // When QTabBarPrivate::dragInProgress is true, and that the mouse is outside of a region
-    // around the QTabBar, we will consider the user wants to drag that QDockWidget away from this
-    // tab area.
+    int hovered = tabAt( e->pos() );
+    if ( hovered != hovered_index )
+    {
+        auto btnA = tabButton( hovered_index, QTabBar::RightSide );
+        auto btnB = tabButton( hovered, QTabBar::RightSide );
 
+        hovered_index = hovered;
+
+        if ( btnA ) btnA->setVisible( false );
+        if ( btnB ) btnB->setVisible( true );
+    }
+
     QTabBarPrivate *d = static_cast<QTabBarPrivate*>(d_ptr.data());
-    if (!draggingDock && (mainWindow->dockOptions() & QMainWindow::GroupedDragging)) {
+    if ( !draggingDock && d->validIndex( d->pressedIndex) ) 
+    {
         int offset = QApplication::startDragDistance() + 1;
-        offset *= 3;
-        QRect r = rect().adjusted(-offset, -offset, offset, offset);
-        if (d->dragInProgress && !r.contains(e->pos()) && d->validIndex(d->pressedIndex)) {
+        QRect r = tabRect( d->pressedIndex ).adjusted(-offset, -offset, offset, offset);
+
+        //
+        // garry: if we drag the tab outside of the tab
+        //
+        if ( !r.contains(e->pos()) )  
+        {
             QMainWindowLayout* mlayout = qt_mainwindow_layout(mainWindow);
             QDockAreaLayoutInfo *info = mlayout->dockInfo(this);
             Q_ASSERT(info);
             int idx = info->tabIndexToListIndex(d->pressedIndex);
+            
             const QDockAreaLayoutItem &item = info->item_list.at(idx);
-            if (item.widgetItem
-                    && (draggingDock = qobject_cast<QDockWidget *>(item.widgetItem->widget()))) {
+
+            //
+            // And we're somehow assciated with a valid DockWidget
+            //
+            if (item.widgetItem && (draggingDock = qobject_cast<QDockWidget *>(item.widgetItem->widget()))) 
+            {
                 // We should drag this QDockWidget away by unpluging it.
                 // First cancel the QTabBar's internal move
                 d->moveTabFinished(d->pressedIndex);
                 d->pressedIndex = -1;
+
                 if (d->movingTab)
-                    d->movingTab->setVisible(false);
+                    d->movingTab->setVisible( false );
+
                 d->dragStartPosition = QPoint();
 
-                // Then starts the drag using QDockWidgetPrivate's API
+                //
+                // Start the drag using QDockWidgetPrivate's API
+                //
                 QDockWidgetPrivate *dockPriv = static_cast<QDockWidgetPrivate *>(QObjectPrivate::get(draggingDock));
                 QDockWidgetLayout *dwlayout = static_cast<QDockWidgetLayout *>(draggingDock->layout());
                 dockPriv->initDrag(dwlayout->titleArea().center(), true);
                 dockPriv->startDrag(false);
+
                 if (dockPriv->state)
                     dockPriv->state->ctrlDrag = e->modifiers() & Qt::ControlModifier;
             }
         }
     }
 
-    if (draggingDock) {
+    if (draggingDock) 
+    {
         QDockWidgetPrivate *dockPriv = static_cast<QDockWidgetPrivate *>(QObjectPrivate::get(draggingDock));
-        if (dockPriv->state && dockPriv->state->dragging) {
+        if (dockPriv->state && dockPriv->state->dragging) 
+        {
             QPoint pos = e->globalPos() - dockPriv->state->pressPos;
             draggingDock->move(pos);
             // move will call QMainWindowLayout::hover
@@ -1770,6 +1808,7 @@
 
     if (e->type() != QEvent::ToolTip)
         return QTabBar::event(e);
+        
     QSize size = this->size();
     QSize hint = sizeHint();
     if (shape() == QTabBar::RoundedWest || shape() == QTabBar::RoundedEast) {
@@ -1794,14 +1833,16 @@
     }
 
     QTabBar *result = nullptr;
-    if (!unusedTabBars.isEmpty()) {
+    if (!unusedTabBars.isEmpty()) 
+    {
         result = unusedTabBars.takeLast();
-    } else {
+    } else 
+    {
         result = new QMainWindowTabBar(static_cast<QMainWindow *>(parentWidget()));
-        result->setDrawBase(true);
+        result->setDrawBase(false);
         result->setElideMode(Qt::ElideRight);
         result->setDocumentMode(_documentMode);
-        result->setMovable(true);
+        result->setMovable(false);
         connect(result, SIGNAL(currentChanged(int)), this, SLOT(tabChanged()));
         connect(result, &QTabBar::tabMoved, this, &QMainWindowLayout::tabMoved);
     }
@@ -2157,7 +2198,7 @@
     if (qobject_cast<QDockWidget*>(widget) != 0) {
         QDockWidgetLayout *layout = qobject_cast<QDockWidgetLayout*>(widget->layout());
         if (layout->nativeWindowDeco()) {
-            globalRect.adjust(0, layout->titleHeight(), 0, 0);
+            globalRect.adjust(0, 0, 0, 0);
         } else {
             int fw = widget->style()->pixelMetric(QStyle::PM_DockWidgetFrameWidth, nullptr, widget);
             globalRect.adjust(-fw, -fw, fw, fw);
--- a/src/widgets/widgets/qtabwidget.cpp	2026-08-22 17:50:07
+++ b/src/widgets/widgets/qtabwidget.cpp	2026-08-22 17:36:46
@@ -1408,7 +1408,6 @@
     Q_D(QTabWidget);
     d->tabs->setDocumentMode(enabled);
     d->tabs->setExpanding(!enabled);
-    d->tabs->setDrawBase(enabled);
     setUpLayout();
 }
 
--- a/src/corelib/global/qmalloc.cpp	2026-08-22 17:50:07
+++ b/src/corelib/global/qmalloc.cpp	2026-08-22 17:36:45
@@ -56,20 +56,37 @@
 Q_CORE_EXPORT void *qRealloc(void *ptr, size_t size) Q_ALLOC_SIZE(2);
 #endif
 
+// VALVE-BEGIN
+typedef void *( *QAlloc_t )( size_t );
+typedef void ( *QFree_t )( void * );
+typedef void *( *QRealloc_t)(void *, size_t );
 
+// These are defined in valve/qtbase/src/valve_memoverride/valve_memoverride.cpp
+// and hooked up to our tier0 allocator when present.
+extern QAlloc_t s_pAlloc;
+extern QRealloc_t s_pRealloc;
+extern QFree_t s_pFree;
+// VALVE-END
+
 void *qMalloc(size_t size)
 {
-    return ::malloc(size);
+// VALVE-BEGIN
+    return s_pAlloc(size);
+// VALVE-END
 }
 
 void qFree(void *ptr)
 {
-    ::free(ptr);
+// VALVE-BEGIN
+    return s_pFree(ptr);
+// VALVE-END
 }
 
 void *qRealloc(void *ptr, size_t size)
 {
-    return ::realloc(ptr, size);
+// VALVE-BEGIN
+    return s_pRealloc(ptr, size);
+// VALVE-END
 }
 
 void *qMallocAligned(size_t size, size_t alignment)
@@ -131,7 +148,10 @@
     if (!ptr)
         return;
     void **ptr2 = static_cast<void **>(ptr);
-    free(ptr2[-1]);
+
+// VALVE-BEGIN
+	s_pFree(ptr2[-1]);
+// VALVE-END
 }
 
 QT_END_NAMESPACE
--- a/qmake/Makefile.unix	2026-08-22 17:57:40
+++ b/qmake/Makefile.unix	2026-08-22 17:58:05
@@ -34,7 +34,8 @@
 	qlocale.o qlocale_tools.o qmap.o qregexp.o qringbuffer.o \
 	qstringbuilder.o qstring.o qstringlist.o qversionnumber.o \
 	qvsnprintf.o qxmlstream.o qxmlutils.o \
-	$(QTOBJS) $(QTOBJS2)
+	$(QTOBJS) $(QTOBJS2) \
+	valve_memoverride.o
 # QTOBJS and QTOBJS2 are populated by Makefile.unix.* as for QTSRC (see below).
 # Note: qlibraryinfo.o omitted deliberately - see below.
 
@@ -129,7 +130,8 @@
 	   $(SOURCE_PATH)/src/corelib/tools/qmap.cpp \
 	   $(SOURCE_PATH)/src/corelib/tools/qringbuffer.cpp \
 	   $(SOURCE_PATH)/src/corelib/tools/qversionnumber.cpp \
-	   $(QTSRCS) $(QTSRCS2)
+	   $(QTSRCS) $(QTSRCS2) \
+	   $(SOURCE_PATH)/src/valve_memoverride/valve_memoverride.cpp
 # QTSRCS and QTSRCS2 come from Makefile.unix.* (concatenated with this
 # by configure); QTSRCS2 may include *.mm entries on macOS.
 # See below for special handling of qlibraryinfo.cpp
@@ -496,6 +498,9 @@
 	$(CXX) -c -o $@ $(CXXFLAGS) $<
 
 qtextstream.o: $(SOURCE_PATH)/src/corelib/serialization/qtextstream.cpp
+	$(CXX) -c -o $@ $(CXXFLAGS) $<
+
+valve_memoverride.o: $(SOURCE_PATH)/src/valve_memoverride/valve_memoverride.cpp
 	$(CXX) -c -o $@ $(CXXFLAGS) $<
 
 # DO NOT DELETE THIS LINE -- make depend depends on it
--- /dev/null
+++ b/src/valve_memoverride/valve_memoverride.cpp
@@ -0,0 +1,25 @@
+#include <dlfcn.h>
+#include <cstdlib>
+
+typedef void *(*QAlloc_t)(size_t);
+typedef void (*QFree_t)(void *);
+typedef void *(*QRealloc_t)(void *, size_t);
+
+QAlloc_t s_pAlloc = &malloc;
+QRealloc_t s_pRealloc = &realloc;
+QFree_t s_pFree = &free;
+
+typedef void (*QGetCustomAllocators_t)(QAlloc_t *, QFree_t *, QRealloc_t *);
+
+__attribute__((constructor(101)))
+static void qt_force_install_custom_allocators()
+{
+    QGetCustomAllocators_t p = (QGetCustomAllocators_t)dlsym(RTLD_DEFAULT, "Qt_GetCustomAllocators");
+    if (p) p(&s_pAlloc, &s_pFree, &s_pRealloc);
+}
+
+void *operator new[](size_t size) { return s_pAlloc(size); }
+void *operator new(size_t size) { return s_pAlloc(size); }
+void operator delete[](void *ptr) { s_pFree(ptr); }
+void operator delete[](void *ptr, size_t) { s_pFree(ptr); }
+void operator delete(void *ptr) { s_pFree(ptr); }
+void operator delete(void *ptr, size_t) { s_pFree(ptr); }
--- /dev/null
+++ b/mkspecs/features/sbox_memoverride.prf
@@ -0,0 +1 @@
+!staticlib: SOURCES += $$PWD/../../src/valve_memoverride/valve_memoverride.cpp
--- a/.qmake.conf
+++ b/.qmake.conf
@@ -7,3 +7,4 @@
 QT_BUILD_TREE = $$shadowed($$PWD)
 
 MODULE_VERSION = 5.15.19
+load(sbox_memoverride)
